test-app/tests/integration/components/headless-form-validation-test.gts

```diff
@@ -830,4 +845,778 @@ module('Integration Component HeadlessForm > Validation', function (hooks) {
+
+        await render(<template>
+          <HeadlessForm @data={{data}} @validateOn="change" as |form|>
+            <form.field
+              @name="firstName"
+              @validate={{validateCallback}}
+              as |field|
+            >
+              <field.label>First Name</field.label>
+              <field.input data-test-first-name />
+            </form.field>
+            <form.field @name="lastName" as |field|>
+              <field.label>Last Name</field.label>
+              <field.input data-test-last-name />
+            </form.field>
+            <button type="submit" data-test-submit>Submit</button>
+          </HeadlessForm>
+        </template>);
+
+        await fillIn('[data-test-first-name]', 'Foo');
+
+        assert.true(
+          validateCallback.calledWith('Foo', 'firstName', {
+            ...data,
+            firstName: 'Foo',
+          }),
+          '@validate is called with form data'
+        );
+
+        await click('[data-test-submit]');
+
+        assert.true(
+          validateCallback.calledTwice,
+          '@validate is called again on submit'
+        );
+      });
+
+      test('validation errors are exposed as field.errors on change', async function (assert) {
+        const data: TestFormData = { firstName: 'Tony', lastName: 'Foo' };
+
+        await render(<template>
+          <HeadlessForm
+            @data={{data}}
+            @validateOn="change"
+            @validate={{validateFormCallbackSync}}
+            as |form|
+          >
+            <form.field @name="firstName" as |field|>
+              <field.label>First Name</field.label>
+              <field.input data-test-first-name />
+              <field.errors data-test-first-name-errors />
+            </form.field>
+            <form.field @name="lastName" as |field|>
+              <field.label>Last Name</field.label>
+              <field.input data-test-last-name />
+              <field.errors data-test-last-name-errors />
+            </form.field>
+            <button type="submit" data-test-submit>Submit</button>
+          </HeadlessForm>
+        </template>);
+
+        assert
+          .dom('[data-test-first-name-errors]')
+          .doesNotExist(
+            'validation errors are not rendered before validation happens'
+          );
+        assert
+          .dom('[data-test-last-name-errors]')
+          .doesNotExist(
+            'validation errors are not rendered before validation happens'
+          );
+
+        await fillIn('[data-test-first-name]', 'Foo');
+
+        assert
+          .dom('[data-test-first-name-errors]')
+          .exists(
+            { count: 1 },
+            'validation errors appear on blur when validation fails'
+          );
+        assert
+          .dom('[data-test-last-name-errors]')
+          .doesNotExist(
+            'validation errors are not rendered for untouched fields'
+          );
+      });
+    });
+  });
+
+  module(`@revalidateOn`, function () {
```

Comment:
I have mixed feelings about this, because it is hard to draw the boundaries here for what should go into what testing file, as all these things belong together and can be combined. For example I don't have test coverage here for all the combinations, like `@validateOn="blur" @revalidateOn="change"`. But this would be a good and valid test scenario, and into which test file should be put it then?

Also we will probably add additional validation-related features, like this `captureEvents` modifier (aka "option 4"), which we might want to test with a particular `@revalidateOn` config. So should this then be in the `validation` file, the `revalidate` file or its own? 

Idk, like I am not opposed to this, as I said I am just unsure how to draw the boundaries and not cause more confusion than we wanted to prevent by splitting this up...

Reply:
Sounds good to me, let's leave it as is